cpp/ql/src/experimental/Security/CWE/CWE-754/ImproperCheckReturnValueScanf.ql

```diff
@@ -0,0 +1,103 @@
+/**
+ * @name Improper check return value scanf
+ * @description Using a function call without the ability to evaluate the correctness of the work can lead to unexpected results.
+ * @kind problem
+ * @id cpp/improper-check-return-value-scanf
+ * @problem.severity warning
+ * @precision medium
+ * @tags correctness
+ *       security
+ *       external/cwe/cwe-754
+ *       external/cwe/cwe-908
+ */
+
+import cpp
+import semmle.code.cpp.commons.Exclusions
+
+/** Returns the position of the first argument being filled. */
+int posArgumentInFunctionCall(FunctionCall fc) {
+  (
+    fc.getTarget().hasGlobalOrStdName(["scanf", "scanf_s"]) and
+    result = 1
+    or
+    fc.getTarget().hasGlobalOrStdName(["fscanf", "sscanf", "fscanf_s", "sscanf_s"]) and
+    result = 2
+  )
+}
+
+/** Holds if a function argument was not initialized but used after the call. */
+predicate argumentIsNotInitializedAndIsUsed(Variable vt, FunctionCall fc) {
+  exists(Expr e0 |
+    // Fillable argument was not initialized.
+    vt instanceof LocalScopeVariable and
+    not vt.getAnAssignment().getASuccessor+() = fc and
+    (
+      not vt.hasInitializer()
+      or
+      exists(Expr e1, Variable v1 |
+        e1 = vt.getInitializer().getExpr() and
+        v1 = e1.(AddressOfExpr).getOperand().(VariableAccess).getTarget() and
+        (
+          not v1.hasInitializer() and
+          not v1.getAnAssignment().getASuccessor+() = fc
+        )
+      )
+    ) and
+    not exists(AssignExpr ae |
+      ae.getLValue() = vt.getAnAccess().getParent() and
+      ae.getASuccessor+() = fc
+    ) and
+    not exists(FunctionCall f0 |
+      f0.getAnArgument().getAChild() = vt.getAnAccess() and
+      f0.getASuccessor+() = fc
+    ) and
+    // After the call, the completed arguments are assigned or returned as the result of the operation of the upper function.
+    fc.getASuccessor+() = e0 and
+    (
+      (
+        e0.(Assignment).getRValue().(VariableAccess).getTarget() = vt or
+        e0.(Assignment).getRValue().(ArrayExpr).getArrayBase().(VariableAccess).getTarget() = vt
+      )
+      or
+      e0.getEnclosingStmt() instanceof ReturnStmt and
+      e0.(VariableAccess).getTarget() = vt
+      or
+      not exists(Expr e1 |
+        fc.getASuccessor+() = e1 and
+        e1.(VariableAccess).getTarget() = vt
+      )
```

Comment:
What does this do? If I remove this, all tests still pass.

Reply:
Yes, you are right.
this section is for finding cases where the variable is not used inside functions that call 1111.
we removed this case from the test cases, replacing it with a return from the function.
but I would like to point out that I would like to keep this detection zone.

Reply:
Please add a test case.

Reply:
Presumably the test case for this is `functionWork3b`?

Reply:
yes, sorry for not specifying that explicitly.